File: plugins/collector/event/apps/database/elasticsearch.go
package database

import (
	"collector/cache/process"
	"collector/event/apps"
	"path/filepath"
	"regexp"
	"strings"
)

type ElasticSearch struct {
	version    string
	reg        *regexp.Regexp
	regVersion *regexp.Regexp
}

func (ElasticSearch) Name() string { return "elasticsearch" }

func (ElasticSearch) Type() string { return "database" }

func (e ElasticSearch) Version() string { return e.version }

func (e *ElasticSearch) Match(p *process.Process) bool {
	if p.Name != "java" {
		return false
	}
	if e.reg == nil {
		e.reg = regexp.MustCompile(` org\.elasticsearch\.bootstrap\.Elasticsearch( )`)
	}
	if s := e.reg.FindString(p.Argv); s == "" {
		return false
	}
	return true
}

func (e *ElasticSearch) Run(p *process.Process) (mapping map[string]string, err error) {
	e.version = ""
	if e.regVersion == nil {
		e.regVersion = regexp.MustCompile(`elasticsearch-(\d+\.)+(\d+)\.jar`)
	}
	if fds, err := p.Fds(); err == nil {
		for _, fd := range fds {
			if filepath.Ext(fd) != ".jar" {
				continue
			}
			if s := e.regVersion.FindString(filepath.Base(fd)); s != "" {
				e.version = strings.TrimRight(strings.TrimPrefix(s, "elasticsearch-"), ".jar")
			}
		}
	}
	return
}

func init() { apps.Regist(&ElasticSearch{}) }

File: plugins/collector/src/event/apps/database/elasticsearch.rs
use crate::event::apps::{find_version, jar_names_for_pid, AppProc, IApp};
use std::collections::HashMap;

pub struct ElasticSearch {
    version: String,
}

impl ElasticSearch {
    pub fn new() -> Self { ElasticSearch { version: String::new() } }
}

impl IApp for ElasticSearch {
    fn name(&self)     -> &'static str { "elasticsearch" }
    fn app_type(&self) -> &'static str { "database" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool {
        p.name == "java"
            && p.argv.contains("org.elasticsearch.bootstrap.Elasticsearch")
    }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        self.version.clear();
        // "elasticsearch-8.4.0.jar"
        for jar in jar_names_for_pid(p.pid) {
            if jar.starts_with("elasticsearch-") && jar.ends_with(".jar") {
                let mid = jar.trim_start_matches("elasticsearch-").trim_end_matches(".jar");
                let ver = find_version(mid);
                if !ver.is_empty() { self.version = ver; break; }
            }
        }
        Ok(HashMap::new())
    }
}
